homeassistant/components/mqtt/trigger.py

```diff
@@ -40,43 +49,35 @@ async def async_attach_trigger(
     trigger_info: TriggerInfo,
 ) -> CALLBACK_TYPE:
     """Listen for state changes based on configuration."""
-    trigger_data = trigger_info["trigger_data"]
-    topic = config[CONF_TOPIC]
-    wanted_payload = config.get(CONF_PAYLOAD)
-    value_template = config.get(CONF_VALUE_TEMPLATE)
-    encoding = config[CONF_ENCODING] or None
-    qos = config[CONF_QOS]
+    trigger_data: dict[str, Any] = dict(trigger_info["trigger_data"])
+    command_template: Callable[..., PublishPayloadType] = MqttCommandTemplate(
+        config.get(CONF_PAYLOAD), hass=hass
+    ).async_render
+    value_template: Callable[..., ReceivePayloadType] = MqttValueTemplate(
+        config.get(CONF_VALUE_TEMPLATE), hass=hass
+    ).async_render_with_possible_json_value
+    encoding: str | None = config[CONF_ENCODING] or None
+    qos: int = config[CONF_QOS]
     job = HassJob(action)
-    variables = None
+    variables: Mapping[str, Any] | None = None
     if trigger_info:
         variables = trigger_info.get("variables")
 
-    template.attach(hass, wanted_payload)
-    if wanted_payload:
-        wanted_payload = wanted_payload.async_render(
-            variables, limited=True, parse_result=False
-        )
+    wanted_payload: PublishPayloadType = command_template(variables=variables)
 
-    template.attach(hass, topic)
-    if isinstance(topic, template.Template):
-        topic = topic.async_render(variables, limited=True, parse_result=False)
-        topic = mqtt.util.valid_subscribe_topic(topic)
-
-    template.attach(hass, value_template)
+    topic_template: Template = config[CONF_TOPIC]
+    topic_template.hass = hass
+    topic: str = mqtt.util.valid_subscribe_topic(
+        topic_template.async_render(variables, limited=True, parse_result=False)
+    )
 
     @callback
-    def mqtt_automation_listener(mqttmsg):
+    def mqtt_automation_listener(mqttmsg: ReceiveMessage) -> None:
         """Listen for MQTT messages."""
-        payload = mqttmsg.payload
-
-        if value_template is not None:
-            payload = value_template.async_render_with_possible_json_value(
-                payload,
-                error_value=None,
-            )
+        payload: ReceivePayloadType = value_template(mqttmsg.payload, default=None)
```

Comment:
This is also a sentinel usecase:
See https://github.com/home-assistant/core/pull/81553